### DISCUSSION TITLE
When or which version the Legacy and low-level API is actually disabled?

### DISCUSSION THREAD

**Discussion by a1exlism** [openssl/openssl#21832] [Q&A] [answered]
Hello Community,
Refer to [Migration Guidance](https://www.openssl.org/docs/man3.0/man7/migration_guide.html#Low-Level-APIs), the low-level API will be deprecated in future versions, but does not explicit what the future means.
My question is, is the `future versions` cross the major version? as `v3.0.x` and `v3.1.x`. It's important for us to implement this in development, Thanks.

**Comment by t8m** [comment, marked as answer]
By Legacy we usually mean the legacy crypto algorithms in the legacy provider - these are not going away in the foreseeable future.

As for the deprecated low-level/engine and other calls they can be and will be removed in a future major version. However it is not yet decided whether it will be the 4.0 or 5.0 version and when these versions are going to be released. There is not even a rough timeline for these releases.

There is only this guidance from the versioning policy:
**No existing public interface can be removed until its replacement has been in place in an LTS stable release. The original interface must remain in at least one supported release for at least 5 years after it is first documented as deprecated in any release.**

As 3.0 where these deprecations happened is LTS, it means the APIs deprecated in 3.0 can be removed in 4.0.

**Reply by huiyuexu** [reply]
As https://openssl-library.org/roadmap/index.html, OpenSSL 4.0 will be released in Apr 2026.  Would the deprecated low-level/engine and other calls be removed in OpenSSL 4.0? @t8m

**Reply by t8m** [reply]
It is not yet decided which deprecated calls & functionality be removed in OpenSSL 4.0. Any currently deprecated functions can be removed but maybe some of them will be still kept - we will need to discuss & decide this through the Business and Technical Advisory Committees of the Foundation and the Corporation.

**Reply by t-j-h** [reply]
Note that engines have been noted numerous times as will-be-removed in OpenSSL 4.0 so that is likely to happen - but as @t8m notes - all of these are subject to community feedback via the BACs and TACs.
